File: src/main/java/insogroup/utils/utils/IOC/strategies/SimpleIOCStrategy.java
package insogroup.utils.utils.IOC.strategies;

import insogroup.utils.exceptions.ClassExtractException;
import insogroup.utils.utils.ClassExtractor;
import insogroup.utils.utils.IOC.exceptions.ResolveStrategyException;
import insogroup.utils.utils.IOC.interfaces.IOCStrategy;

public class SimpleIOCStrategy implements IOCStrategy {

    protected String classPath;

    /**
     * @param classPath package.name for using with @{@link ClassExtractor} in the future
     */
    public SimpleIOCStrategy(String classPath) throws ClassExtractException {
        ClassExtractor.getClass(classPath);
        this.classPath = classPath;
    }

    /**
     * @param args Args for using in strategy
     * @return new instance
     * @throws ResolveStrategyException Throw when object with this classPath and args can't be instantiated
     */
    @Override
    public Object getInstance(Object... args) throws ResolveStrategyException {
        try {
            Object instance = ClassExtractor.extract(classPath, args);
            return instance;
        } catch (ClassExtractException e) {
            throw new ResolveStrategyException("Can't resolve instance with this classPath or args", e);
        }
    }
}

File: src/main/kotlin/com/github/insanusmokrassar/utils/IOC/strategies/SimpleIOCStrategy.kt
package com.github.insanusmokrassar.utils.IOC.strategies

import com.github.insanusmokrassar.utils.ClassExtractor.exceptions.ClassExtractException
import com.github.insanusmokrassar.utils.ClassExtractor.ClassExtractor
import com.github.insanusmokrassar.utils.IOC.exceptions.ResolveStrategyException
import com.github.insanusmokrassar.utils.IOC.interfaces.IOCStrategy

class SimpleIOCStrategy(protected var classPath: String) : IOCStrategy {


    /**
     * @param classPath package.name for using with @[ClassExtractor] in the future
     */
    init {
        ClassExtractor.getClass<Any>(classPath)
    }

    /**
     * @param args Args for using in strategy
     * *
     * @return new instance
     * *
     * @throws ResolveStrategyException Throw when object with this classPath and args can't be instantiated
     */
    @Throws(ResolveStrategyException::class)
    override fun getInstance(vararg args: Any): Any {
        try {
            val instance = ClassExtractor.extract<Any>(classPath, *args)
            return instance
        } catch (e: ClassExtractException) {
            throw ResolveStrategyException("Can't resolve instance with this classPath or args", e)
        }

    }
}
